langchain4j-coherence/src/main/java/dev/langchain4j/store/embedding/coherence/CoherenceEmbeddingStore.java

```diff
@@ -0,0 +1,532 @@
+     *         call {@link Embedding#normalize()} on embeddings when adding or searching
+     */
+    public boolean isForceNormalize() {
+        return forceNormalize;
+    }
+
+/**
+     * Add an embedding to the repository.
+     *
+     * @param id         the id of the {@link Embedding}
+     * @param embedding  the {@link Embedding} to add
+     * @param segment    an optional {@link TextSegment} to add with the embedding
+     */
+    private void addInternal(DocumentChunk.Id id, Embedding embedding, TextSegment segment) {
+        Map<DocumentChunk.Id, DocumentChunk> map = new HashMap<>();
+        map.put(id, createChunk(embedding, segment));
+        documentChunks.putAll(map);
+    }
+
+    /**
+     * Add multiple {@link Embedding} instances to the repository.
+     *
+     * @param ids         the list of identifiers to use for each of the {@link Embedding}
+     * @param embeddings  the {@link Embedding} to add
+     * @param segments    an optional list of {@link TextSegment} to add with the embeddings
+     */
+    private void addAllInternal(List<DocumentChunk.Id> ids, List<Embedding> embeddings, List<TextSegment> segments) {
+        if (isNullOrEmpty(ids) || isNullOrEmpty(embeddings)) {
+            Logger.info("Skipped adding empty embeddings");
+            return;
+        }
+
+        boolean hasEmbedded = segments != null && !segments.isEmpty();
+
+        ensureTrue(ids.size() == embeddings.size(), "ids size is not equal to embeddings size");
+        if (hasEmbedded) {
+            ensureTrue(embeddings.size() == segments.size(), "embeddings size is not equal to embedded size");
+        }
+
+        Map<DocumentChunk.Id, DocumentChunk> map = new HashMap<>();
+        for (int i = 0; i < embeddings.size(); i++) {
+            Embedding embedding = embeddings.get(i);
+            TextSegment segment = hasEmbedded ? segments.get(i) : null;
+            map.put(ids.get(i), createChunk(embedding, segment));
+        }
+        documentChunks.putAll(map);
+    }
+
+    /**
+     * Convert a {@link QueryResult} into an {@link EmbeddingMatch}.
+     *
+     * @param score  the relevance score
+     * @param id     the chunk ID
+     * @param chunk  the matched chunk
+     *
+     * @return an {@link EmbeddingMatch} created from the {@link QueryResult}
+     */
+    private EmbeddingMatch<TextSegment> embeddingMatch(double score, DocumentChunk.Id id, DocumentChunk chunk) {
+        String key = id.toString();
+        String text = chunk.text();
+        TextSegment segment = text == null ? null : new TextSegment(chunk.text(), getMetadata(chunk.metadata()));
+        Vector<float[]> vector = chunk.vector();
+        Embedding embedding = vector == null ? null : new Embedding(vector.get());
+
+        return new EmbeddingMatch<>(score, key, embedding, segment);
+    }
+
+    private static Metadata getMetadata(Map<String, Object> mapMetadata) {
+        Metadata metadata = new Metadata();
```

Comment:
```suggestion
        mapMetadata.entrySet().removeIf(entry -> entry.getValue() == null);
        return Metadata.from(mapMetadata);
```

`Metadata#add` is deprecated, so can we just filter and construct it using `Metadata#from`?